packages/block-library/src/table-of-contents/index.php

```diff
@@ -0,0 +1,174 @@
+			/**
+			 * Calculate the last index before the next iteration with the same level (siblings).
+			 * Then, use this index to slice the array for use in recursion.
+			 * This prevents duplicate nodes.
+			 */
+			$end = $total;
+			for ( $i = $offset; $i < $total; $i++ ) {
+				if ( $headings[ $i ]['level'] === $heading['level'] ) {
+					$end = $i;
+					break;
+				}
+			}
+
+			$tree[] = array(
+				'heading'  => $heading,
+				'children' => block_core_table_of_contents_build_headings_tree( array_slice( $headings, $offset, $end - 1 ) ),
+			);
+
+		} else {
+			// No child heading levels.
+			$tree[] = array(
+				'heading' => $heading,
+			);
+		}
+	}
+
+	return $tree;
+}
+
+/**
+ * Build a list of Table of Content items.
+ *
+ * @since 6.4.0
+ *
+ * @param array $tree An array of nested headings.
+ * @return string $list A list of Table of Content items.
+ */
+function block_core_table_of_contents_build_list( $tree ) {
+	$list      = '';
+	$permalink = get_permalink();
+
+	foreach ( $tree as $item ) {
+		$heading  = $item['heading'];
+		$children = isset( $item['children'] ) ? '<ol>' . block_core_table_of_contents_build_list( $item['children'] ) . '</ol>' : '';
+
+		if ( ! empty( $heading['link'] ) ) {
+			$pagelink = ! empty( $heading['page'] ) ? add_query_arg( 'page', $heading['page'], $permalink ) : $permalink;
+			$content  = '<a class="wp-block-table-of-contents__entry" href="' . esc_url( $pagelink . $heading['link'] ) . '">' . esc_html( $heading['content'] ) . '</a>';
+		} else {
+			$content = '<span class=wp-block-table-of-contents__entry>' . esc_html( $heading['content'] ) . '</span>';
+		}
+
+		$list .= '<li>' . $content . $children . '</li>';
+	}
+
+	return $list;
+}
+
+/**
+ * Renders the `core/table-of-contents` block on the server.
+ *
+ * @since 6.4.0
+ *
+ * @param array    $attributes Block attributes.
+ * @param string   $content    Block default content.
+ * @param WP_Block $block      Block instance.
+ *
+ * @return string Returns the HTML representing the table of contents.
+ */
+function render_block_core_table_of_contents( $attributes, $content, $block ) {
+	// Fallback to static content when it exists. The post hasn't been updated to use headings via post-meta.
+	if ( trim( $content ) !== '' ) {
+		return $content;
+	}
+
+	// Bail out early if the post ID is not set for some reason.
+	if ( empty( $block->context['postId'] ) ) {
+		return '';
+	}
+
+	if ( post_password_required( $block->context['postId'] ) ) {
+		return '';
+	}
+
+	$headings = get_post_meta( $block->context['postId'], 'table_of_contents', true );
+
+	if ( ! $headings ) {
+		return '';
+	}
+
+	$headings = json_decode( $headings, true );
+
+	if ( ! is_array( $headings ) || count( $headings ) === 0 ) {
+		return '';
+	}
+
+	$tree = block_core_table_of_contents_build_headings_tree( $headings );
+
+	return sprintf(
+		'<nav %1$s><ol>%2$s</ol></nav>',
+		get_block_wrapper_attributes(),
+		block_core_table_of_contents_build_list( $tree )
+	);
+}
```

Comment:
There are various ways to edit the content of a post through the API, like using the REST API for example (POST to a `/wp/v2/posts/<id>`), or a raw `wp_update_post()` call from a plugin. There the user can provide `content` (or `post_content`) with changes to the headings, and from what I understand, these changes won't be reflected in the relevant post meta field. Is that something that's being considered?

Reply:
The same argument could be made for the Footnotes block. It's hard to do anything for similar updates when observers are on the client side.

Reply:
I definitely don't want to block this PR, however, to be fair, this isn't a good reason IMHO. This only means that we need to consider the same scenarios for the Footnotes block. 

Updating a post through:

* REST API
* WP CLI
* `save_post` hook and `wp_update_post()` and family 
* SQL query  

is a valid post-updating scenario and we should not be skipping any of them. Otherwise, we'll the metadata can end up being obsolete, and I'd bet we don't want that.

Reply:
IMO ToC is takes post content as input only? Why does it even need post meta? Is that just for caching? Maybe invalidate the cache on save?

Reply:
Data is derived (from blocks), saved in post meta, and then used by render callback. Currently, deriving data is much easier in the editor.

Another option is to parse post content on the server and generate ToC on the fly. The editor will use a simple state for rendering preview.

The problem with server-side parsing is that heading block attributes like `content` and `id` aren't serialized; we can't use `parse_blocks` to derive ToC headings.

I guess we could use the new HTML API. @dmsnell, would the API be suitable for this use case?

P.S. I'm discounting `RegEx` as an option since it's hard to maintain and unreliable when it comes to parsing HTML.

Reply:
What if there are other blocks with headings? Maybe use the tag processor to just get all the headings?

Reply:
> What if there are other blocks with headings? Maybe use the tag processor to just get all the headings?

I don't have a historical context on this one, but I think it was a safer assumption that Heading blocks will always be a part of the content structure while other headings that are part of blocks aren't. cc @ZebulanStanphill

Reply:
> that would make it much harder to render them outside of the post content in a template

why is this? if the table of contents is a block it seems like we could quickly grab it with `parse_blocks()` and handle it elsewhere. that should be faster than making a DB call.

> would the [HTML] API be suitable for this use case?

yes and no! I'd love to explore this together. it depends on what we're wanting to do, specifically. if it's simply to discover every `H1` through `H6` and link to them in order, that should be easy. if it's looking more in depth to determine what should be a header, who knows. let's figure it out.

the _HTML Processor_ is not ready yet for it because it's currently lacking support for list elements and other common tags, and it bails when it encounters unsupported elements, but the _Tag Processor_ can help.

the Tag Processor could even build a map of all `id` values that are linked within the page with a corresponding `<a href="#{id}">` if it mattered.

Reply:
> > What if there are other blocks with headings? Maybe use the tag processor to just get all the headings?
> 
> I don't have a historical context on this one, but I think it was a safer assumption that Heading blocks will always be a part of the content structure while other headings that are part of blocks aren't. cc @ZebulanStanphill

Actually, here is the reasoning that led to the current implementation lacking support for headings outside of Heading blocks: https://github.com/WordPress/gutenberg/issues/29560

I think the idea was that eventually we'd add a dedicated API for handling document heading structure, which would be abstracted from the actual blocks/content used. Combined with the complexities of parsing HTML (my implementation back then relied on libxml), it seemed easier to just support Heading blocks at the time.

Of course, the abstract headings API remains an unimplemented concept, and now we have the HTML Tag Processor to make parsing HTML easier.

With regard to "what should count as a heading", I think it's fairly simple, conceptually: any h2-h6 tag inside the post content is a heading, unless it is inside a nested `<article>` or `<aside>` . This would align the most with actual HTML semantics. (As a side note, an `<h1>` in the content would be ignored since that should only ever be used for the post title, which would be excluded from the table of contents anyway.)

Perhaps a slightly more difficult question would be "how do we determine where the post content is". Presumably, we could just look for the parent `<article>` to determine the boundaries of where to look for heading elements, or just look at the entire rendered document if on an old theme that doesn't use HTML5.

Reply:
The Tag Processor should be able to find all H1-H6 elements, but it's not technically capable of knowing if they are inside a nested article or aside. it can be used to approximate that though in the meantime until the HTML Processor is ready for that.

places where it could get off track are not only when we have missing closing tags, but also when we encounter boundaries like the TEMPLATE element, tables, and others, which implicitly close tags that might still be open.